solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -246,6 +244,90 @@ public boolean canBeRemoved() {
+     * Gets the DocCollection (state) of the collection which the corresponding watch last observed
+     *
+     * @param collection the collection name to get DocCollection on
+     * @return The last observed DocCollection(state). if null, that means there's no such
+     *     collection.
+     */
+    private DocCollection getDocCollection(String collection) {
+      DocCollectionWatch<DocCollectionWatcher> watch = get(collection);
+      return watch != null ? watch.currentDoc : null;
+    }
+
+    /**
+     * Gets the active collections (collections that exist) being watched
+     *
+     * @return an immutable set of active collection names
+     */
+    private Set<String> activeCollections() {
+      return this.entrySet().stream()
+          .filter(
+              (Entry<String, DocCollectionWatch<DocCollectionWatcher>> entry) ->
+                  entry.getValue().currentDoc != null)
+          .map(Entry::getKey)
+          .collect(Collectors.toUnmodifiableSet());
+    }
+
+    /**
+     * Updates the latest observed DocCollection (state) of the {@link DocCollectionWatch} if the
+     * collection is being watched
+     *
+     * @param collection the collection name
+     * @param newState the new DocCollection (state) observed
+     * @return whether an active watch exists for such collection
+     */
+    private boolean updateDocCollection(String collection, DocCollection newState) {
+      DocCollectionWatch<DocCollectionWatcher> watch = get(collection);
+      if (watch != null) {
+        DocCollection oldState = watch.currentDoc;
+        if (oldState == null && newState == null) {
+          // OK, the collection not yet exist in ZK
+        } else if (oldState == null) {
+          if (log.isDebugEnabled()) {
+            log.debug("Add data for [{}] ver [{}]", collection, newState.getZNodeVersion());
+          }
+          watch.currentDoc = newState;
+        } else if (newState == null) {
+          log.debug("Removing cached collection state for [{}]", collection);
+          watch.currentDoc = null;
+        } else { // both new and old states are non-null
+          int oldCVersion =
+              oldState.getPerReplicaStates() == null ? -1 : oldState.getPerReplicaStates().cversion;
+          int newCVersion =
+              newState.getPerReplicaStates() == null ? -1 : newState.getPerReplicaStates().cversion;
+          if (oldState.getZNodeVersion() < newState.getZNodeVersion()
+              || oldCVersion < newCVersion) {
+            watch.currentDoc = newState;
+            if (log.isDebugEnabled()) {
+              log.debug(
+                  "Updating data for [{}] from [{}] to [{}]",
+                  collection,
+                  oldState.getZNodeVersion(),
+                  newState.getZNodeVersion());
+            }
+          }
+        }
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  private static class DocCollectionWatch<T> extends CollectionWatch<T> {
```

Comment:
If this is only ever used with `T = DocCollectionWatcher` then we might as well declare that explicitly and save ourselves typing on the declarations? There might be other usages that I'm not seeing though. I don't think this is something that's worth trying to make generic (yet?)
```suggestion
  private static class DocCollectionWatch extends CollectionWatch<DocCollectionWatcher> {
```

Reply:
Agreed! especially that it has a `private DocCollection currentState;` which is pretty specific and not generic anyway